### PR TITLE
Improve stream extractor tests & various fixes

### PR DESCRIPTION
**Pull request by Stypox**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).

This PR makes all stream extractor tests consistent by having them all extend the same default base class. I took inspiration from @mauriciocolli's work in #296. ~~Note that I didn't add tests for metadata properties, that will fit in another PR after both this pr and #306 are merged.~~

List of test improvements:
- Add base classes for stream extractor tests with the same structure as in #296, and extend them in each and every StreamExtractorTest. Some methods are not abstract and they provide a default value; this does not imply risks (i.e. forgetting to change a default) since if a value is different from the default one the tests will fail (not silently)
- Thumbnail urls and uploader avatar urls are only `assertSecureUrl`, but not compared, since they could change I think. In MediaCCC, where they were being tested before, I left those tests there untouched.
- Peertube has many description-only tests; I didn't create a whole test class for every one of them, but left them untouched, as a full test is not really needed imo.
- Rename `*StreamExtractorDefaultTest` into `*StreamExtractorTest` for services when there is only one file containing stream tests (that's obviously the default one ;-) )
- Remove assert checking whether the upload date of list items is in the past. Actually, this does not have to hold true: youtube premieres turn up in search results even though they are in the future.
- Add `assertAtLeast` method to utils
- Optimize imports

List of fixes (taken from commit messages):
- [YouTube] Fix `NullPointerException` on obtaining error message when there is none: return null instead, as per javadoc
- [YouTube] Return 0 when there is no timestamp in a video url, not `-2` (that internally indicates that the regex failed to match), as per javadoc
- [YouTube] Fix frame extraction for livestreams: another json field is used in that case. Also use saved playerResponse instead of parsing the same json once again.
- [SoundCloud] Return empty stream list instead of null in getVideoStreams() (best-practice). Also replace every instance of `return new ArrayList<>();` with `return Collections.emptyList();`
- [PeerTube] Return empty audio stream list instead of null in `getAudioStreams()`
- [PeerTube] Fix link handler inconsistency, that used to provide API links to the user instead of the human-readable ones. Also improve stream and channel link handler tests.
- [PeerTube] Parse timestamp from url (previously unimplemented)
- [MediaCCC] Fix link handler inconsistency, that used to provide API links to the user instead of the human-readable ones. Use regex instead of substrings.
- [MediaCCC] Add tests for stream and conference link handlers.
- [MediaCCC] Implement tag metadata extraction (unrelated to the other fixes, will probably be overwritten by #306)
- [MediaCCC] Return null instead of empty items collector at `getRelatedStreams()`, as per javadoc
- [MediaCCC] Remove useless overridden method `getOriginalUrl()` that does the same as the base one
- [MediaCCC] Return empty video-only stream list instead of null in `getVideoOnlyStreams()`

### CODE REVIEW

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/linkHandler/MediaCCCStreamLinkHandlerFactory.java [hunk 1]
```diff
@@ -9,10 +8,12 @@
...
+    public static final String VIDEO_API_ENDPOINT = "https://api.media.ccc.de/public/events/";
+    private static final String VIDEO_PATH = "https://media.ccc.de/v/";
 
     @Override
     public String getId(String urlString) throws ParsingException {
-        if (urlString.startsWith("https://api.media.ccc.de/public/events/") &&
+        if (urlString.startsWith(VIDEO_API_ENDPOINT) &&
                 !urlString.contains("?q=")) {
             return urlString.substring(39); //remove api…/public/events part
         }
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/linkHandler/MediaCCCStreamLinkHandlerFactory.java:19]: Why are you using a regex for conferences while using `substring()` for API URLs here and using `URL#getPath()` + `substring()` for non-API URLs here?

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeChannelLinkHandlerFactory.java [hunk 2]
```diff
@@ -31,7 +32,7 @@ public String getUrl(String id, List<String> contentFilters, String searchFilter
     @Override
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeChannelLinkHandlerFactory.java:32]: Please also handle the API URLs in `getId()` here

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeStreamLinkHandlerFactory.java [hunk 3]
```diff
@@ -27,7 +28,7 @@ public String getUrl(String id) {
 
     @Override
     public String getUrl(String id, String baseUrl) {
-        return baseUrl + VIDEO_ENDPOINT + id;
+        return baseUrl + VIDEO_PATH + id;
     }
 
     @Override
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeStreamLinkHandlerFactory.java:34]: Please also handle the API URLs in `getId()` here

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java [hunk 4]
```diff
@@ -258,7 +258,14 @@ public long getLength() throws ParsingException {
      */
     @Override
     public long getTimeStamp() throws ParsingException {
-        return getTimestampSeconds("((#|&|\\?)t=\\d{0,3}h?\\d{0,3}m?\\d{1,3}s?)");
+        long timestamp = getTimestampSeconds("((#|&|\\?)t=\\d{0,3}h?\\d{0,3}m?\\d{1,3}s?)");
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java:261]: Could be `final`

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java [hunk 5]
```diff
@@ -579,7 +586,7 @@ public String getErrorMessage() {
         try {
             return getTextFromObject(initialAjaxJson.getObject(2).getObject("playerResponse").getObject("playabilityStatus")
                     .getObject("errorScreen").getObject("playerErrorMessageRenderer").getObject("reason"));
-        } catch (ParsingException e) {
+        } catch (Exception e) {
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java:589]: Why was this changed?

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    @Test
+    @Override
+    public void testThumbnailUrl() throws Exception {
+        assertIsSecureUrl(extractor().getThumbnailUrl());
+    }
+
+    @Test
+    @Override
+    public void testDescription() throws Exception {
+        Description description = extractor().getDescription();
```
Comment (wb9688) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:94]: Could be `final`, just like a lot of variables in this class

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java (hunk 5 above)
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java:589]: So that it also catches NullPointerExceptions, which are generated when there is no error message. In that case `null` should be returned

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java (hunk 6 above)
Comment (Stypox) [reply extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:94]: Ok, will fix this in all modified files

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/linkHandler/MediaCCCStreamLinkHandlerFactory.java (hunk 1 above)
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/linkHandler/MediaCCCStreamLinkHandlerFactory.java:19]: I left everything as it was before, I only replaced the hardcoded url. Will fix this

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeStreamLinkHandlerFactory.java (hunk 3 above)
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeStreamLinkHandlerFactory.java:34]: Oh right, for some reason I thought they were already handled in some strange way

Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeStreamLinkHandlerFactory.java:34]: Those are already handled by the regex: `/videos/(watch/)?([^/?&#]*)`
Both "framatube.org/api/v1/videos/id" and "framatube.org/videos/id" match it and the second group is obtained correctly

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeChannelLinkHandlerFactory.java (hunk 2 above)
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/linkHandler/PeertubeChannelLinkHandlerFactory.java:32]: Fixed while rebasing with the latest changes about accounts and channels

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java (hunk 5 above)
Comment (B0pol) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java:589]: Then why aren't you catching `(ParsingException | NullPointerException e)`?

**Comment by Stypox** [comment]
Ok, I fixed everything that was suggested plus an oversight when testing video streams, check commit messages for more info. @B0pol @wb9688

**Comment by wb9688** [comment]
@Stypox: It's rebase time! Yes, _again_. Anyway, I'll try to take an actual look at it tomorrow.

**Comment by Stypox** [comment]
@wb9688 rebased and squashed some commits (in particular, I squashed all commits related to tests into only one, while living fixes into their separate commits). I learned how to rebase interactively :-D

**Comment by Stypox** [comment]
I rebased

**Comment by Stypox** [comment]
Rebased

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+        videoStreams.addAll(videoOnlyStreams);
+
+        if (expectedHasVideoStreams()) {
+            assertFalse(videoStreams.isEmpty());
+
+            for (VideoStream stream : videoStreams) {
+                assertIsSecureUrl(stream.getUrl());
+                assertFalse(stream.getResolution().isEmpty());
+
+                int formatId = stream.getFormatId();
```
Comment (wb9688) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:210]: `final`

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+        final List<AudioStream> audioStreams = extractor().getAudioStreams();
+        assertNotNull(audioStreams);
+
+        if (expectedHasAudioStreams()) {
+            assertFalse(audioStreams.isEmpty());
+
+            for (AudioStream stream : audioStreams) {
+                assertIsSecureUrl(stream.getUrl());
+
+                int formatId = stream.getFormatId();
```
Comment (wb9688) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:231]: `final`

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+            }
+        } else {
+            assertTrue(audioStreams.isEmpty());
+        }
+    }
+
+    @Test
+    @Override
+    public void testSubtitles() throws Exception {
+        List<SubtitlesStream> subtitles = extractor().getSubtitlesDefault();
```
Comment (wb9688) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:243]: `final`

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+        assertNotNull(subtitles);
+
+        if (expectedHasSubtitles()) {
+            assertFalse(subtitles.isEmpty());
+
+            for (SubtitlesStream stream : subtitles) {
+                assertIsSecureUrl(stream.getUrl());
+
+                int formatId = stream.getFormatId();
```
Comment (wb9688) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:252]: `final`

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+                assertTrue("format id does not fit an audio stream: " + formatId,
+                        0x1000 <= formatId && formatId < 0x10000);
+            }
+        } else {
+            assertTrue(subtitles.isEmpty());
+
+            MediaFormat[] formats = {MediaFormat.VTT, MediaFormat.TTML, MediaFormat.TRANSCRIPT1,
```
Comment (wb9688) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:259]: `final`

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+                    MediaFormat.TRANSCRIPT2, MediaFormat.TRANSCRIPT3, MediaFormat.SRT};
+            for (MediaFormat format : formats) {
+                subtitles = extractor().getSubtitles(format);
```
Comment (wb9688) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:262]: Make this a variable that's declared here, because this is imho unclear

**Comment by Stypox** [comment]
@wb9688 fixed the failing test (for some reason SoundCloud started serving a date as GMT, causing a 2h difference between the expected parsed value and the actual one) and the inconsistencies you have found. Implemented metadata tests (currently only used in PeerTube, waiting for #306) and sub-channel tests. Rebased. ;-)

**Comment by Stypox** [comment]
Oh, apparently tarvis is provided with dates in a different format. What the heck?

**Comment by Stypox** [comment]
I thought the problem lied in PeerTube and SoundCloud date parsers, but actually this problem had to do with Youtube: the upload date wasn't being parsed as a GMT date, but using the device timezone, thus leading to inconsistencies. Tests were also wrong, because they tested dates using the devide timezone (I did try to prevent this, but I put the correct statement on the wrong line ;-/) Now everything seems to work, both on my end and on tarvis. It's incredible how 1085 tests succeed ;-D

**Comment by wb9688** [comment]
Could you rebase?

**Comment by Stypox** [comment]
Done. I'll soon also push a test for the DashMpd url that you recently fixed

**Comment by Stypox** [comment]
Done

**Comment by Stypox** [comment]
Merged with dev and fixed PewDiePie test

**Comment by opusforlife2** [comment]
> fixed PewDiePie test

He needs all the help he can get.

**Comment by Stypox** [comment]
I rebased. There are many failing tests, but all seem to be unrelated from the changes here, since they are caused by "YouTube did not provide player config even after three attempts", PeerTube accounts/channels not providing related items, SoundCloud channels & playlists with some outdated data, youtube channels with no Videos tab, YouTube comments extractor (all not touched by this PR).

**Review by TobiGr — APPROVED** [review]
Looks good to me!
We could make a bunch of vars and arguments final, but that can wait

**Comment by Stypox** [comment]
@TobiGr done ;-)

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+
+        if (expectedHasAudioStreams()) {
+            assertFalse(audioStreams.isEmpty());
+
+            for (final AudioStream stream : audioStreams) {
+                assertIsSecureUrl(stream.getUrl());
+
+                final int formatId = stream.getFormatId();
+                assertTrue("format id does not fit an audio stream: " + formatId,
+                        0x100 <= formatId && formatId < 0x1000);
```
Comment (B0pol) [inline extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:274]: Could you explain why 0x100 <= formatId < 0x1000 and also above (video streams) or below (subtitles) and add comments for it?

**Review by B0pol — CHANGES_REQUESTED** [review]

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/extractors/MediaCCCStreamExtractor.java [hunk 14]
```diff
@@ -216,7 +219,7 @@ public StreamType getStreamType() {
 
     @Override
     public StreamInfoItemsCollector getRelatedStreams() {
-        return new StreamInfoItemsCollector(getServiceId());
+        return null;
```
[the remark this answers is not in the document]
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/extractors/MediaCCCStreamExtractor.java:222]: This was done as per the documentation in the base `getRelatedStreams()`, and also to prevent creating new objects at random. I tested in NewPipe and it works as expected (there are `null` checks), but the code which takes care of it being `null` actually lies [in the extractor](https://github.com/TeamNewPipe/NewPipeExtractor/blob/030465b5d4d8c738a42cfc52bd1d78f54b0f6bd4/extractor/src/main/java/org/schabi/newpipe/extractor/utils/ExtractorHelper.java#L33), where the collector is turned into a list. I now added @Nullable to all of the functions, as it was missing.

File: extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java (hunk 13 above)
Comment (Stypox) [reply extractor/src/test/java/org/schabi/newpipe/extractor/services/DefaultStreamExtractorTest.java:274]: Ok

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/extractors/MediaCCCStreamExtractor.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/media_ccc/extractors/MediaCCCStreamExtractor.java:222]: Oh, and also, YouTube already returns `null` for age restricted videos.

**Comment by Stypox** [comment]
@B0pol done everything

**Comment by Stypox** [comment]
Testing apk: [app-debug.zip](https://github.com/TeamNewPipe/NewPipeExtractor/files/5434271/app-debug.zip)

**Review by B0pol — APPROVED** [review]
